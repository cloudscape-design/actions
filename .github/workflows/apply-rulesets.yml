name: Apply Github rulesets

on:
  push:
    branches:
      - main
  workflow_dispatch:

permissions:
  contents: read

jobs:
  applyRulesets:
    name: Apply Github rulesets
    runs-on: ubuntu-latest
    env:
      GH_ORG_TOKEN: ${{ secrets.GH_ORG_TOKEN }}
    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Setup environment
        uses: actions/setup-node@v4
        with:
          node-version: 20

      - name: Install dependencies
        run: npm install

      - name: Build the package
        run: npm run build

      - name: Apply the rulesets
        run: node lib/apply-rulesets/index.js
