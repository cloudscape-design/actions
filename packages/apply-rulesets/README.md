# Apply Github rulesets

Sub-module to apply Github org rulesets.

For modifying the Github checks, change the content in `src/constants.ts` file.

## Deploy changes

The changes will be deployed automatically after merging to `main`.
