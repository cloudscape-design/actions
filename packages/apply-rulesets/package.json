{
  "name": "@cloudscape-design/apply-rulesets",
  "version": "1.0.0",
  "main": "dist/index.js",
  "type": "module",
  "scripts": {
    "build": "tsc",
    "test": "vitest run"
  },
  "devDependencies": {
    "@tsconfig/node-ts": "^23.6.1",
    "@tsconfig/node20": "^20.1.6",
    "@types/node": "^20.19.9",
    "typescript": "^5.8.3",
    "vitest": "^3.2.4"
  },
  "dependencies": {
    "@octokit/rest": "^22.0.0"
  }
}
