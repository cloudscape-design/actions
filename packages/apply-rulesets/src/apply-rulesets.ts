import { BASIC_CHECKS, CODECOV_CHECKS, CODECOV_REPOS, DRY_RUN_CHECKS, DRY_RUN_REPOS } from './constants.js';
import { createOrUpdateRuleset } from './create-or-update-ruleset.js';

export async function applyRulesets() {
  await createOrUpdateRuleset({
    // We'll roll out repositories slowly over time to avoid operational load.
    // As we roll out to a repo, the branch protection rules will be deleted for that repo.
    // The value will be eventually replaced with ['~ALL'] at the end of the roll-out
    repos: ['actions', 'test-utils', 'demos', 'components'],

    name: 'Basic rulesets',
    rules: [
      {
        type: 'pull_request',
        parameters: {
          require_code_owner_review: false,
          require_last_push_approval: true,
          required_approving_review_count: 1,
          required_review_thread_resolution: false,
          allowed_merge_methods: ['merge'],
          dismiss_stale_reviews_on_push: true,
        },
      },
      {
        type: 'required_status_checks',
        parameters: {
          strict_required_status_checks_policy: true,
          required_status_checks: BASIC_CHECKS.map(context => ({ context })),
        },
      },
    ],
  });

  await createOrUpdateRuleset({
    name: 'Code-cov rulesets',
    repos: CODECOV_REPOS,
    rules: [
      {
        type: 'required_status_checks',
        parameters: {
          strict_required_status_checks_policy: true,
          required_status_checks: CODECOV_CHECKS.map(context => ({ context })),
        },
      },
    ],
  });

  await createOrUpdateRuleset({
    name: 'Dry-run rulesets',
    repos: DRY_RUN_REPOS,
    rules: [
      {
        type: 'required_status_checks',
        parameters: {
          strict_required_status_checks_policy: true,
          required_status_checks: DRY_RUN_CHECKS.map(context => ({ context })),
        },
      },
    ],
  });
}
