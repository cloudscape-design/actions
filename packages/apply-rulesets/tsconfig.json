{
  "extends": ["@tsconfig/node20", "@tsconfig/node-ts"],
  "compilerOptions": {
    "outDir": "../../lib/apply-rulesets"
  },
  "include": ["src"],
  "exclude": ["node_modules"]
}
