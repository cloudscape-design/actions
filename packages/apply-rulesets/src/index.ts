import { applyRulesets } from './apply-rulesets.js';

try {
  await applyRulesets();
  console.log('Rulesets applied successfully.');
} catch (error) {
  console.log('Failed to apply rulesets.');
  throw error;
}
