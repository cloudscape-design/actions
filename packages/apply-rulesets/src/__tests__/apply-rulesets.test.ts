import { vi, describe, it, expect } from 'vitest';
import { applyRulesets } from '../apply-rulesets.ts';

const mocks = vi.hoisted(() => ({
  getOrgRulesets: vi.fn().mockResolvedValue({ data: [] }),
  updateOrgRuleset: vi.fn(),
  createOrgRuleset: vi.fn(),
  getRepoRulesets: vi.fn().mockResolvedValue({ data: [] }),
  updateRepoRuleset: vi.fn(),
  createRepoRuleset: vi.fn(),
}));
vi.mock('@octokit/rest', () => ({
  Octokit: vi.fn().mockReturnValue({
    rest: {
      repos: {
        getOrgRulesets: mocks.getOrgRulesets,
        updateOrgRuleset: mocks.updateOrgRuleset,
        createOrgRuleset: mocks.createOrgRuleset,
        getRepoRulesets: mocks.getRepoRulesets,
        updateRepoRuleset: mocks.updateRepoRuleset,
        createRepoRuleset: mocks.createRepoRuleset,
      },
    },
  }),
}));

describe(applyRulesets.name, () => {
  it('applies all of the rulesets', async () => {
    await applyRulesets();
    expect(mocks.createOrgRuleset.mock.calls).toMatchSnapshot();
    expect(mocks.createRepoRuleset.mock.calls).toMatchSnapshot();
  });
});
