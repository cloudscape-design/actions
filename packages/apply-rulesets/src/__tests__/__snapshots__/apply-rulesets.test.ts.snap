// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`applyRulesets > applies all of the rulesets 1`] = `
[
  [
    {
      "conditions": {
        "ref_name": {
          "exclude": [],
          "include": [
            "~DEFAULT_BRANCH",
          ],
        },
        "repository_name": {
          "exclude": [],
          "include": [
            "actions",
            "test-utils",
            "demos",
            "components",
          ],
        },
      },
      "enforcement": "active",
      "name": "Basic rulesets",
      "org": "cloudscape-design",
      "rules": [
        {
          "type": "required_linear_history",
        },
        {
          "parameters": {
            "allowed_merge_methods": [
              "squash",
            ],
            "dismiss_stale_reviews_on_push": true,
            "require_code_owner_review": false,
            "require_last_push_approval": true,
            "required_approving_review_count": 1,
            "required_review_thread_resolution": false,
          },
          "type": "pull_request",
        },
        {
          "parameters": {
            "required_status_checks": [
              {
                "context": "build / build",
              },
              {
                "context": "build / codeql",
              },
              {
                "context": "build / git-secrets",
              },
            ],
            "strict_required_status_checks_policy": true,
          },
          "type": "required_status_checks",
        },
      ],
      "target": "branch",
    },
  ],
  [
    {
      "conditions": {
        "ref_name": {
          "exclude": [],
          "include": [
            "~DEFAULT_BRANCH",
          ],
        },
        "repository_name": {
          "exclude": [],
          "include": [
            "components",
          ],
        },
      },
      "enforcement": "active",
      "name": "Code-cov rulesets",
      "org": "cloudscape-design",
      "rules": [
        {
          "parameters": {
            "required_status_checks": [
              {
                "context": "codecov/patch",
              },
              {
                "context": "codecov/project",
              },
            ],
            "strict_required_status_checks_policy": true,
          },
          "type": "required_status_checks",
        },
      ],
      "target": "branch",
    },
  ],
  [
    {
      "conditions": {
        "ref_name": {
          "exclude": [],
          "include": [
            "~DEFAULT_BRANCH",
          ],
        },
        "repository_name": {
          "exclude": [],
          "include": [
            "test-utils",
            "components",
          ],
        },
      },
      "enforcement": "active",
      "name": "Dry-run rulesets",
      "org": "cloudscape-design",
      "rules": [
        {
          "parameters": {
            "required_status_checks": [
              {
                "context": "dry-run / Build board components",
              },
              {
                "context": "dry-run / Build browser-test-tools",
              },
              {
                "context": "dry-run / Build chat components",
              },
              {
                "context": "dry-run / Build code view components",
              },
              {
                "context": "dry-run / Build collection-hooks",
              },
              {
                "context": "dry-run / Build component-toolkit",
              },
              {
                "context": "dry-run / Build components",
              },
              {
                "context": "dry-run / Build documenter",
              },
              {
                "context": "dry-run / Build global-styles",
              },
              {
                "context": "dry-run / Build jest-preset",
              },
              {
                "context": "dry-run / Build test-utils",
              },
              {
                "context": "dry-run / Build theming-core",
              },
              {
                "context": "dry-run / Components accessibility tests shards (1)",
              },
              {
                "context": "dry-run / Components accessibility tests shards (2)",
              },
              {
                "context": "dry-run / Components accessibility tests shards (3)",
              },
              {
                "context": "dry-run / Components accessibility tests shards (4)",
              },
              {
                "context": "dry-run / Components accessibility tests shards (5)",
              },
              {
                "context": "dry-run / Components accessibility tests shards (6)",
              },
              {
                "context": "dry-run / Components integration tests shards (1)",
              },
              {
                "context": "dry-run / Components integration tests shards (2)",
              },
              {
                "context": "dry-run / Components integration tests shards (3)",
              },
              {
                "context": "dry-run / Components integration tests shards (4)",
              },
              {
                "context": "dry-run / Components motion tests",
              },
              {
                "context": "dry-run / Components unit tests",
              },
              {
                "context": "dry-run / Create build tree",
              },
              {
                "context": "dry-run / Demos tests",
              },
              {
                "context": "dry-run / Validate PR title",
              },
            ],
            "strict_required_status_checks_policy": true,
          },
          "type": "required_status_checks",
        },
      ],
      "target": "branch",
    },
  ],
]
`;

exports[`applyRulesets > applies all of the rulesets 2`] = `
[
  [
    {
      "conditions": {
        "ref_name": {
          "exclude": [],
          "include": [
            "~DEFAULT_BRANCH",
          ],
        },
      },
      "enforcement": "active",
      "name": "Merge-queue ruleset",
      "owner": "cloudscape-design",
      "repo": "components",
      "rules": [
        {
          "parameters": {
            "check_response_timeout_minutes": 60,
            "grouping_strategy": "ALLGREEN",
            "max_entries_to_build": 5,
            "max_entries_to_merge": 1,
            "merge_method": "SQUASH",
            "min_entries_to_merge": 1,
            "min_entries_to_merge_wait_minutes": 5,
          },
          "type": "merge_queue",
        },
      ],
      "target": "branch",
    },
  ],
]
`;
